File: 01_projects/code/python-playwright/cbaps-dawms/basic-pom/cbaps/pages/funding_lines_page.py
"""
FundingLinesPage: Add and manage funding lines for a requisition
Allows adding multiple funding lines and continuing to routing
"""

import logging
from playwright.sync_api import Page, Locator
from cbaps.library.playwright_manager import PlaywrightManager

logger = logging.getLogger(__name__)


class FundingLinesPage:
    """
    FundingLinesPage handles funding line management.
    This is the Python equivalent of the Java FundingLinesPage.
    """
    
    def __init__(self, page: Page, pwm: PlaywrightManager):
        """
        Initialize FundingLinesPage
        
        Args:
            page: Playwright Page object
            pwm: PlaywrightManager instance
        """
        self.page = page
        self.pwm = pwm
        
        # Locators
        self.add_line_button: Locator = page.locator("button:has-text('Add Line')")
        self.amount_input: Locator = page.locator("#fundAmount")
        self.save_line_button: Locator = page.locator("button:has-text('Save')")
        self.continue_to_routing_button: Locator = page.locator(
            "button:has-text('Continue to Routing')"
        )
        
        # Stability anchor: wait for page to load
        pwm.wait_visible("text=Funding Lines")
        logger.info("Funding Lines Page loaded successfully.")
    
    def add_funding_line(self, amount: str) -> 'FundingLinesPage':
        """
        Add a funding line with specified amount
        
        Args:
            amount: Funding amount
            
        Returns:
            self for method chaining
        """
        self.pwm.click(self.add_line_button)
        logger.info("Clicked 'Add Line' button.")
        
        self.pwm.type(self.amount_input, amount)
        logger.info(f"Entered funding amount: {amount}")
        
        self.pwm.click(self.save_line_button)
        logger.info("Clicked 'Save' button for funding line.")
        
        return self
    
    def continue_to_routing(self) -> 'RoutingApprovalPage':
        """
        Continue to Routing/Approval step
        
        Returns:
            RoutingApprovalPage instance
        """
        from cbaps.pages.routing_approval_page import RoutingApprovalPage
        
        self.pwm.click(self.continue_to_routing_button)
        logger.info("Clicked 'Continue to Routing' button.")
        return RoutingApprovalPage(self.page, self.pwm)

File: 01_projects/code/java-playwright/cbaps-dawms/basic-pom/src/test/java/com/cbaps/pages/FundingLinesPage.java
package com.cbaps.pages;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.cbaps.library.PlaywrightManager;

/**
 * FundingLinesPage: Add and manage funding lines for a requisition
 * Allows adding multiple funding lines and continuing to routing
 */
public class FundingLinesPage {

	private static final Logger log = LoggerFactory.getLogger(FundingLinesPage.class);
	private final Page page;
	private final PlaywrightManager pwm;

	private final Locator addLineButton;
	private final Locator amountInput;
	private final Locator saveLineButton;
	private final Locator continueToRoutingButton;

	public FundingLinesPage(Page page, PlaywrightManager pwm) {
		this.page = page;
		this.pwm = pwm;

		// Initialize locators
		this.addLineButton = page.locator("button:has-text('Add Line')");
		this.amountInput = page.locator("#fundAmount");
		this.saveLineButton = page.locator("button:has-text('Save')");
		this.continueToRoutingButton = page.locator("button:has-text('Continue to Routing')");

		// Stability anchor: wait for page to load
		pwm.waitVisible("text=Funding Lines");
		log.info("Funding Lines Page loaded successfully.");
	}

	/**
	 * Add a funding line with specified amount
	 * @param amount - Funding amount
	 * @return this (for method chaining)
	 */
	public FundingLinesPage addFundingLine(String amount) {
		pwm.click(addLineButton);
		log.info("Clicked 'Add Line' button.");
		
		pwm.type(amountInput, amount);
		log.info("Entered funding amount: " + amount);
		
		pwm.click(saveLineButton);
		log.info("Clicked 'Save' button for funding line.");
		
		return this;
	}

	/**
	 * Continue to Routing/Approval step
	 * @return RoutingApprovalPage
	 */
	public RoutingApprovalPage continueToRouting() {
		pwm.click(continueToRoutingButton);
		log.info("Clicked 'Continue to Routing' button.");
		return new RoutingApprovalPage(page, pwm);
	}
}
